### PR TITLE
backport: bitcoin#17725, #18304, #18392, #18430, #18438, #18441, #18477, #18562, #18569, #18683, #18899, #18912, #19159, #19669, partial #14794 - asan/valgrind CI changes

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
This PR is the next batch of bitcoin backports, mostly related to CI, address sanitizer and valgrind.

## What was done?
- bitcoin/bitcoin#17725
- bitcoin/bitcoin#18304
- bitcoin/bitcoin#18392
- bitcoin/bitcoin#19159
- bitcoin/bitcoin#18899
- bitcoin/bitcoin#18430
- bitcoin/bitcoin#18438
- bitcoin/bitcoin#18441
- bitcoin/bitcoin#18477
- bitcoin/bitcoin#18562
- bitcoin/bitcoin#18569
- bitcoin/bitcoin#18683
- bitcoin/bitcoin#18912
- bitcoin/bitcoin#19669
- **partial** bitcoin/bitcoin#14794

There are discovered 2 issues that won't let to enable Asan CI and valgrind CI at the moment.

**1. Asan (address sanitizer) generates a lot of failures like that both CI and locally:**
```
********* Finished testing of TrafficGraphDataTests *********
~BitcoinApplication : Stopping thread
~BitcoinApplication : Stopped thread
==28475==WARNING: invalid path to external symbolizer!
==28475==WARNING: Failed to use and restart external symbolizer!
=================================================================
==28475==ERROR: LeakSanitizer: detected memory leaks
Direct leak of 520 byte(s) in 13 object(s) allocated from:
    #0 0x55fc08e2353d  (/builds/dashpay/dash/build-ci/dashcore-linux64_asan/src/qt/test/test_dash-qt+0x779153d)
    #1 0x7f7a7724bd85  (<unknown module>)
Direct leak of 432 byte(s) in 2 object(s) allocated from:
    #0 0x55fc08e2353d  (/builds/dashpay/dash/build-ci/dashcore-linux64_asan/src/qt/test/test_dash-qt+0x779153d)
    #1 0x55fc0dc63b63  (/builds/dashpay/dash/build-ci/dashcore-linux64_asan/src/qt/test/test_dash-qt+0xc5d1b63)
    #2 0x8604597ff0e8feff  (<unknown module>)
Direct leak of 176 byte(s) in 2 object(s) allocated from:
    #0 0x55fc08e52cad  (/builds/dashpay/dash/build-ci/dashcore-linux64_asan/src/qt/test/test_dash-qt+0x77c0cad)
    #1 0x55fc0d822372  (/builds/dashpay/dash/build-ci/dashcore-linux64_asan/src/qt/test/test_dash-qt+0xc190372)
Direct leak of 66 byte(s) in 3 object(s) allocated from:
    #0 0x55fc08e2353d  (/builds/dashpay/dash/build-ci/dashcore-linux64_asan/src/qt/test/test_dash-qt+0x779153d)
    #1 0x7f7a77243633  (<unknown module>)
Direct leak of 24 byte(s) in 1 object(s) allocated from:
    #0 0x55fc08e2353d  (/builds/dashpay/dash/build-ci/dashcore-linux64_asan/src/qt/test/test_dash-qt+0x779153d)
    #1 0x7f7a77245449  (<unknown module>)
```
Here's run on CI: https://gitlab.com/dashpay/dash/-/pipelines/883649210
I tried different ways to add more debug infos or check what is these addresses but I haven't localized issue at the moment.
So, ASAN build is disabled at the moment.

**2.  valgrind fails for each usage of std::optional when it passes as an argument in function params.**
```
==6024== Conditional jump or move depends on uninitialised value(s) 
==6024==    at 0xF6C314: CWallet::ScanForWalletTransactions(uint256 const&, int, std::optional<int>, WalletRescanReserver const&, bool) (wallet.cpp:1976)
==6024==    by 0x8245D8: wallet_tests::scan_for_wallet_transactions::test_method() (wallet_tests.cpp:106)
==6024==    by 0x8267B4: wallet_tests::scan_for_wallet_transactions_invoker() (wallet_tests.cpp:87)
==6024==    by 0x115D4F1: boost::detail::function::function_obj_invoker0<boost::detail::forward, int>::invoke(boost::detail::function::function_buffer&) (in DASH/src/test/test_dash)
==6024==    by 0x115BB9C: boost::execution_monitor::catch_signals(boost::function<int ()> const&) (in DASH/src/test/test_dash)
==6024==    by 0x115BC24: boost::execution_monitor::execute(boost::function<int ()> const&) (in DASH/src/test/test_dash)
==6024==    by 0x115BD04: boost::execution_monitor::vexecute(boost::function<void ()> const&) (in DASH/src/test/test_dash)
==6024==    by 0x1122834: boost::unit_test::unit_test_monitor_t::execute_and_translate(boost::function<void ()> const&, unsigned long) (in DASH/src/test/test_dash)
==6024==    by 0x10FC18F: boost::unit_test::framework::state::execute_test_tree(unsigned long, unsigned long, boost::unit_test::framework::state::random_generator_helper const*) [clone .isra.0] (in DASH/src/test/test_dash)
==6024==    by 0x10FC489: boost::unit_test::framework::state::execute_test_tree(unsigned long, unsigned long, boost::unit_test::framework::state::random_generator_helper const*) [clone .isra.0] (in DASH/src/test/test_dash)
==6024==    by 0x10FC489: boost::unit_test::framework::state::execute_test_tree(unsigned long, unsigned long, boost::unit_test::framework::state::random_generator_helper const*) [clone .isra.0] (in DASH/src/test/test_dash)
==6024==    by 0x11004D6: boost::unit_test::framework::run(unsigned long, bool) (in DASH/src/test/test_dash)
```
There's only one example, but it happens for all other places with std::optional too.
`valgrind` CI is actually removed from bitcoin and super-seeded by address sanitizer, but it doesn't work anyway at the moment for dash's code.

## How Has This Been Tested?
Run functional and unit tests + build locally with valgrind/address sanitizer to see a result.
Based on local runs this PR includes several fixes of UB.

## Breaking Changes
No breaking changes.

## Checklist:
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -289,15 +289,6 @@ linux64_nowallet-build:
   variables:
     BUILD_TARGET: linux64_nowallet
 
-linux64_valgrind-build:
```
[the remark this answers is not in the document]
Comment (knst) [reply .gitlab-ci.yml]: our CI fails anyway for valgrind ATM (see my PR description, seems as a bug with std::optional implementation on valgrind).


let me actually keep everything but comment CI call [that anyway fails] maybe?

[the remark this answers is not in the document]
Comment (UdjinM6) [reply .gitlab-ci.yml]: > let me actually keep everything but comment CI call [that anyway fails] maybe?

agree

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Pull request merged**